utilities/backupable/backupable_db.cc

```diff
@@ -2159,84 +2215,182 @@ Status BackupEngineImpl::BackupMeta::LoadFromFile(
+    }
+    std::string field_name = line.substr(0, space_pos);
+    std::string field_data = line.substr(space_pos + 1);
+    if (field_name == kAppMetaDataFieldName) {
       // app metadata present
-      data.remove_prefix(kMetaDataPrefix.size());
-      bool decode_success = data.DecodeHex(&app_metadata_);
+      bool decode_success = Slice(field_data).DecodeHex(&app_metadata_);
       if (!decode_success) {
         return Status::Corruption(
             "Failed to decode stored hex encoded app metadata");
       }
-      line.clear();
+    } else if (schema_major_version < 2) {
+      return Status::Corruption("Expected number of files or \"" +
+                                kAppMetaDataFieldName + "\" field");
+    } else if (StartsWith(field_name, "ni_")) {
+      // ni_ short for "non-ignorable"
+      return Status::NotSupported("Unrecognized non-ignorable meta field " +
+                                  field_name);
     } else {
-      // process the line below
+      // Warn the first time we see any particular unrecognized meta field
+      if (reported_ignored_fields->insert(field_name + "1").second) {
+        ROCKS_LOG_WARN(info_log, "Ignoring unrecognized backup meta field %s",
+                       field_name.c_str());
+      }
     }
-  } else {
-    line.clear();
-  }
-  uint32_t num_files = UINT32_MAX;
-  if (!line.empty() || backup_meta_reader->ReadLine(&line)) {
-    num_files = static_cast<uint32_t>(strtoul(line.c_str(), nullptr, 10));
   }
   std::vector<std::shared_ptr<FileInfo>> files;
+  bool footer_present = false;
   while (backup_meta_reader->ReadLine(&line)) {
     std::vector<std::string> components = StringSplit(line, ' ');
 
     if (components.size() < 1) {
       return Status::Corruption("Empty line instead of file entry.");
     }
+    if (schema_major_version >= 2 && components.size() == 2 &&
+        components[0] == "//" && components[1] == "footer") {
+      footer_present = true;
+      break;
+    }
 
     const std::string& filename = components[0];
 
-    uint64_t size;
+    uint64_t actual_size;
     const std::shared_ptr<FileInfo> file_info = GetFile(filename);
     if (file_info) {
-      size = file_info->size;
+      actual_size = file_info->size;
     } else {
       std::string abs_path = backup_dir + "/" + filename;
       auto e = abs_path_to_size.find(abs_path);
       if (e == abs_path_to_size.end()) {
         return Status::Corruption("Pathname in meta file not found on disk: " +
                                   abs_path);
       }
-      size = e->second;
+      actual_size = e->second;
     }
 
-    if (components.size() < 3) {
-      return Status::Corruption("File checksum is missing for " + filename +
-                                " in " + meta_filename_);
+    if (schema_major_version >= 2) {
+      if (components.size() % 2 != 1) {
+        return Status::Corruption(
+            "Bad number of line components for file entry.");
+      }
```

Comment:
Is components.size() == 1 valid?

Reply:
Yes, we might have just the file name and no size or checksum. I envision us always including the size when we put the new format into production, but it is useful for testing individual aspects of the old format (size not included; crc32(c) checksum included). Plus it is good to leave options open.

Reply:
By the way, I did not elect to solve the problem of spaces in filenames. That would have been easy with JSON but we don't have an existing escape/unescape functions that escape spaces (e.g. with https://en.wikipedia.org/wiki/Percent-encoding). But this can be added later with something like

    ni::file_name_escape uri_percent